investos/portfolio/risk_model/factor_risk.py

```diff
@@ -1,23 +1,351 @@
+                    temp_df, how="left", on="id"
+                )
+                self._fos_risk_factor_uuids.remove(factor)
+            else:
+                self._df_loadings = pd.merge_asof(
+                    self._df_loadings.sort_values("datetime"),
+                    temp_df.sort_values("datetime"),
+                    on="datetime",
+                    by="id",
+                    direction="forward",
+                )
+        # handling missing data and inf data
+        self._df_loadings = self._df_loadings.replace(
+            [np.inf, -np.inf], np.nan
+        ).dropna()
+        self._df_loadings = self._df_loadings[
+            (self._df_loadings.datetime >= self._end_date)
+            & (self._df_loadings.datetime <= self._start_date)
+        ]
+
+        print("Done pulling data")
+        return
+
+    def _generate_risk_models(self, timestamp="datetime") -> None:
+        """fit a cross sectional factor model and generate factor return covariance and idio returns variance
 
-        self._drop_excluded_assets()
+        Parameters
+        ----------
+        self._df_loadings : DataFrame
+            Linear regression model fits a linear model
+        self._fos_risk_factor_uuids : list
+            the array of X cols (AKA factor columns, or training data columns)
+        self._regress_col : string
+            the y col, by default return_1d
+
+        Returns
+        -------
+        """
+
+        """
+        fit the cross sectional regression
+        """
+        print("\n2. Running Cross Sectional Regression...")
+        self._fos_industry_factors = [
+            col
+            for col in self._df_loadings.columns
+            if col
+            not in ["datetime", "id"]
+            + self._fos_risk_factor_uuids
+            + [self._fos_return_uuid]
+        ]
+        scaler = StandardScaler()
+        self._factor_returns = {}
+        for date in sorted(self._df_loadings[timestamp].unique()):
+            # isolating the data for the current date
+            df_current = self._df_loadings[self._df_loadings[timestamp] == date]
+
+            # winsorzie and standarize risk factors
+            for col in self._fos_risk_factor_uuids:
+                df_current.loc[:, col] = mstats.winsorize(
+                    df_current.loc[:, col], limits=(0.05, 0.95)
+                )
+                df_current.loc[:, col] = scaler.fit_transform(
+                    df_current.loc[:, col].values.reshape(-1, 1)
+                ).flatten()
+            # winsorzie return
+            df_current.loc[:, self._fos_return_uuid] = mstats.winsorize(
+                df_current.loc[:, self._fos_return_uuid], limits=(0.05, 0.95)
+            )
+
+            y, X = (
+                df_current.loc[:, self._fos_return_uuid],
+                df_current.loc[
+                    :, (self._fos_risk_factor_uuids + self._fos_industry_factors)
+                ],
+            )
+            X.loc[:, "const"] = (
```

Comment:
This should come out of PR. Can't have this in OS

Reply:
Thanks for discussing with me. Let's keep it in, but let's only use a global (intercept) factor set to 1 for all assets if add_global_constant_factor = True in `__init__`

Reply:
Instead of calling the factor const, let's call it global_const